### PR TITLE
Use __slots__ for ProtocolElements

### PR DESCRIPTION
**Pull request by Naburimannu**
Replaces `__dict__` with `__slots__` for all ProtocolElement subclasses,
leading to significant space savings. Switches unit test from `__dict__`
to `__slots__`.

Addresses https://github.com/ga4gh/server/issues/92

-------
This cuts memory usage during our benchmark variants search by 250MB.

In master, server_benchmark.py --profile heap shows:
Partition of a set of 1253757 objects. Total size = 402956592 bytes.
```
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0 250400  20 262419200  65 262419200  65 dict of ga4gh._protocol_definitions.GACall
     1 250500  20 70140000  17 332559200  83 dict (no owner)
     2 501502  40 54216624  13 386775824  96 list
     3 250400  20 16025600   4 402801424 100 ga4gh._protocol_definitions.GACall
     4    100   0   104800   0 402906224 100 dict of ga4gh._protocol_definitions.GAVariant
     5    406   0    19440   0 402925664 100 str
     6    139   0    15496   0 402941160 100 unicode
     7    100   0     6400   0 402947560 100 ga4gh._protocol_definitions.GAVariant
     8    200   0     6400   0 402953960 100 long
     9      1   0     1048   0 402955008 100 dict of
                                             ga4gh._protocol_definitions.GASearchVariantsRequest
```

With this patch in place:
Partition of a set of 1003255 objects. Total size = 152460144 bytes.
```
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0 250500  25 70140000  46  70140000  46 dict (no owner)
     1 501502  50 54216624  36 124356624  82 list
     2 250400  25 28044800  18 152401424 100 ga4gh._protocol_definitions.GACall
     3    406   0    19440   0 152420864 100 str
     4    100   0    16000   0 152436864 100 ga4gh._protocol_definitions.GAVariant
     5    139   0    15496   0 152452360 100 unicode
     6    200   0     6400   0 152458760 100 long
     7      2   0      888   0 152459648 100 types.FrameType
     8      3   0      264   0 152459912 100 __builtin__.weakref
     9      1   0      128   0 152460040 100 ga4gh._protocol_definitions.GASearchVariantsRequest
```

### CODE REVIEW

**Comment by dcolligan** [comment]
+1, but looks like you have some pep8 errors

**Comment by kerrydc** [comment]
+1, same as above

**Comment by Naburimannu** [comment]
Ugh, we're running PEP8 on the generated code in _protocol_definitions.py.

**Comment by dcolligan** [comment]
I think that's intentional

**Comment by Naburimannu** [comment]
PEP8 should be happier now. Unfortunately it looks like postMethods isn't stably-ordered; filing an issue to fix that.

**Comment by dcolligan** [comment]
That can be solved with a `self.postSignatures.sort()` at the end of `SchemaGenerator.__init__()`, if you wouldn't mind putting it in there in this PR.

**Comment by dcolligan** [comment]
(Or perhaps there is a deeper issue that is causing this that we want to solve instead of just calling the sort method to fix it... )

File: scripts/generate_schemas.py [hunk 1]
```diff
@@ -131,6 +132,15 @@ def formatRequiredFields(self):
         return string
 
     def writeConstructor(self, outputFile):
+        # Force using slots to avoid the overhead of a dict per object;
+        # when a query returns hundreds of thousands of calls this can
+        # save a hundred megabytes or more.
+        print("    __slots__ = ['",
+              textwrap.fill(
```
Comment (jeromekelleher) [inline scripts/generate_schemas.py]: Nice! Wish I'd known about textwrap earlier. I've filed an issue saying we should update the rest of this script to follow the same pattern.

**Comment by jeromekelleher** [comment]
+1.

This is a huge step forward @Naburimannu, thanks!

**Pull request merged**